### ISSUE TITLE
Functionality With External Depth Maps

### ISSUE THREAD

**Issue by Joeljensen28** [colmap/colmap#2996]
I am trying to make reconstructions of homes, but a lot of the results I've gotten are incomplete, especially in weakly textured areas (which I understand is within the performance expectations). I want to test how monocular depth/normal estimation may improve the reconstruction of textureless areas.

Does COLMAP have the ability to do this? If not, it would be really great if there was a simple way to integrate external depth maps into the COLMAP pipeline.

**Comment by samhubs** [comment]
Great suggestion. I am curious if you are using a RGBD camera. If not, how are you planning to estimate the depth maps to import because if you are doing an image-based depth estimation outside of colmap, wouldn't that struggle with the lack of texture too?

**Comment by Joeljensen28** [comment]
I am not using RGBD, just my plain old phone camera. I am using other open-source ML-based libraries that have pre-trained models, such as Metric3D. I'm certainly not an expert on the subject, but I think it handles them very well (hence why I would like to test it with an SfM/MVS pipeline 🤗).

**Comment by tsattler** [comment]
Colmap does not support using external depth maps during the sparse reconstruction stage.

**Comment by Joeljensen28** [comment]
> Colmap does not support using external depth maps during the sparse reconstruction stage.

How difficult would it be to implement such a thing? I'm a pretty junior programmer, but I'd be thrilled to give it a shot myself.

**Comment by tsattler** [comment]
Strongly depends on how you want to use the depth maps / normal maps. Even if you just use them to initialize 3D point positions for local features, you will still have to handle the fact that the scale of most depth maps is not in any meaningful coordinate system and that you might need to re-scale all depth maps.

**Comment by mattiasmar** [comment]
The RealSense depth cameras give metric depth. This should give us a strong prior to the 3D points' positions.  @tsattler  @sarlinpe Do you know if there is a PR/fork/branch exploring  using this information?

**Comment by sarlinpe** [comment]
We have a working prototype of SfM aided by monocular/sensor depth. We might release it in the coming months. Could you point us to a dataset (with ground truth poses and ideally sensor depth) that captures your use case? ScanNet?

**Comment by mattiasmar** [comment]
No, I don't have such a dataset to share. But I can propose ideas on how to upgrade existing dataset. For example one could apply any of the depth estimation models released the last year and corrupt their estimates to resemble more what one would get from the different kinds of depth sensing devices in the market.  

Would you be interested in collaborating on the prototype of SfM aided by monocular/sensor depth?

**Comment by nlml** [comment]
@sarlinpe  Any updates on this prototype of incorporating sensor depth? If so I would be curious to test it with depth maps generated from an iPhone pro LiDAR scanner.

**Comment by sarlinpe** [comment]
You can check out https://github.com/cvg/mpsfm We've only used it with monocular depth but in principle sensor depth work fine too.

**Comment by ahojnnes** [comment]
Closing: external depth maps functionality answered; pointed to mpsfm.
